tests/middleware/test_http.py

```diff
@@ -0,0 +1,208 @@
+            yield None
+
+    class bMiddleware(HTTPMiddleware):
+        async def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+            scope["state_bar"] = expected_value2
+            response = yield None
+            response.headers["X-State-Foo"] = scope["state_foo"]
+
+    class cMiddleware(HTTPMiddleware):
+        async def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+            response = yield None
+            response.headers["X-State-Bar"] = scope["state_bar"]
+
+    def homepage(request):
+        return PlainTextResponse("OK")
+
+    app = Starlette(
+        routes=[Route("/", homepage)],
+        middleware=[
+            Middleware(aMiddleware),
+            Middleware(bMiddleware),
+            Middleware(cMiddleware),
+        ],
+    )
+
+    client = test_client_factory(app)
+    response = client.get("/")
+    assert response.text == "OK"
+    assert response.headers["X-State-Foo"] == expected_value1
+    assert response.headers["X-State-Bar"] == expected_value2
+
+
+def test_app_middleware_argument(test_client_factory):
+    def homepage(request):
+        return PlainTextResponse("Homepage")
+
+    app = Starlette(
+        routes=[Route("/", homepage)], middleware=[Middleware(CustomMiddleware)]
+    )
+
+    client = test_client_factory(app)
+    response = client.get("/")
+    assert response.headers["Custom-Header"] == "Example"
+
+
+def test_middleware_repr():
+    middleware = Middleware(CustomMiddleware)
+    assert repr(middleware) == "Middleware(CustomMiddleware)"
+
+
+def test_fully_evaluated_response(test_client_factory):
+    # Test for https://github.com/encode/starlette/issues/1022
+    class CustomMiddleware(HTTPMiddleware):
+        async def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+            yield PlainTextResponse("Custom")
+
+    app = Starlette(middleware=[Middleware(CustomMiddleware)])
+
+    client = test_client_factory(app)
+    response = client.get("/does_not_exist")
+    assert response.text == "Custom"
+
+
+def test_exception_on_mounted_apps(test_client_factory):
+    sub_app = Starlette(routes=[Route("/", exc)])
+    app = Starlette(routes=[Mount("/sub", app=sub_app)])
+
+    client = test_client_factory(app)
+    with pytest.raises(Exception) as ctx:
+        client.get("/sub/")
+    assert str(ctx.value) == "Exc"
+
+
+ctxvar: contextvars.ContextVar[str] = contextvars.ContextVar("ctxvar")
+
+
+class CustomMiddlewareWithoutBaseHTTPMiddleware:
+    def __init__(self, app: ASGIApp) -> None:
+        self.app = app
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        ctxvar.set("set by middleware")
+        await self.app(scope, receive, send)
+        assert ctxvar.get() == "set by endpoint"
+
+
+class CustomMiddlewareUsingHTTPMiddleware(HTTPMiddleware):
+    async def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+        ctxvar.set("set by middleware")
+        yield None
```

Comment:
The `yield None` can be avoided by users if they write `-> AsyncGenerator[None, Response]` or `-> AsyncIterable[None]`